Data/Text.hs

```diff
@@ -660,12 +681,91 @@ fastLength (Text arr (Exts.I# off0#) (Exts.I# len0#)) =
+fastNthCodepoint (Exts.I# n0#) (Text arr (Exts.I# off0#) (Exts.I# len0#))
+    -- if n > len there are definitely less than n characters
+    | Exts.isTrue# (n0# >=# len0#) = -1
+    | Exts.isTrue# (n0# <# 0#)     = 0
+    | otherwise = initFastNth off0# (int2Word# n0#) where
+    ba = A.aBA arr
+    end# = off0# +# len0#
+    endVec# = quotInt# end# WORDBYTES
+
+    initFastNth :: Int# -> Word# -> Int
+    initFastNth off# n#
+        | Exts.isTrue# (off# >=# end#)    = -1
+        -- If n less than bytes/word, clean up in the epilogue loop
+        | Exts.isTrue# (n# `leWord#` WORDBYTESW) = endFastNth off# n#
+        -- If word aligned, process in the fast loop
+        | Exts.isTrue# (andI# off# ALIGN_MASK ==# 0#) =
+            -- endFastNth off# n#
+            vecFastNth (quotInt# off# WORDBYTES) n#
+        | otherwise = case indexWord8Array# ba off# of
+            -- isContByte b     = 1|0
+            -- isContByte b - 1 = 0|-1
+            -- therefore, we only subtract 1 from n if it is not a continuation byte
+            b# -> initFastNth (off# +# 1#) (plusWord# n# (minusWord# (isContByte# b#) 1##))
+    
+    -- Count down n one word at a time. Offset is in Word#'s not bytes
+    vecFastNth :: Int# -> Word# -> Int
+    vecFastNth offWord# n#
+        -- Check for end of array, or if there are less than #bytes/word left
+        -- left to find and let endFastNth take care of any cleanup.
+        | Exts.isTrue# ((offWord# >=# endVec#) `orI#` leWord# n# WORDBYTESW) =
+            endFastNth (offWord# *# WORDBYTES) n#
+        | otherwise = vecFastNth
+                (offWord# +# 1#)
+                -- Subtract #bytes/word, add the number of non codepoint start
+                -- chars
+                (plusWord#
+                    (minusWord# n# WORDBYTESW)
+                    (countContBytes# (READWORD ba offWord#)))
+
+    endFastNth :: Int# -> Word# -> Int
+    endFastNth off# n#
+        -- overrun
+        | Exts.isTrue# (off# >=# end#)    = -1
+        -- still more characters to find
+        | Exts.isTrue# (n# `gtWord#` 0##) = let
+            b# = indexWord8Array# ba off#
+            -- isContByte b     = 1|0
+            -- isContByte b - 1 = 0|-1
+            -- therefore, we only subtract 1 from n if it is not a continuation byte
+            in endFastNth (off# +# 1#) (plusWord# n# (minusWord# (isContByte# b#) 1##))
+        -- n == 0
+        | otherwise = cleanEnd off#
+    
+    -- Cleans up the final character, finds the offset to the next
+    -- non-continuation byte
+    cleanEnd :: Int# -> Int
+    cleanEnd off#
+        -- overrun
+        | Exts.isTrue# (off# >=# end#)    = -1
+        -- if it's a continuation byte, search again
+        | Exts.isTrue# (word2Int# (isContByte# (indexWord8Array# ba off#)))
+            = cleanEnd (off# +# 1#)
+        -- If this is not a contiunuation byte, the previous byte is the last one
+        | otherwise                       = Exts.I# off#
```

Comment:
Hi @kuk0 I thought this would be more useful in general, if a user of the function needs to compute off-off0 they can but this code avoids having lots of `(off - off0) + off0` throughout the code in the places where the offset is directly needed.